src/AccountManager.mo

```diff
@@ -491,40 +497,95 @@ module {
       };
     };
 
-    /// Initiates a withdrawal by transferring tokens to another account.
-    /// Returns ICRC1 transaction index and amount of transferred tokens (fee excluded).
-    public func withdraw(to : ICRC1.Account, amount : Nat) : async* WithdrawResponse {
-
-      totalWithdrawn_ += amount;
-
-      let callResult = await* processWithdrawTransfer(to, amount);
+    func handleWithdrawal(
+      from_subaccount : ?Blob,
+      to : ICRC1.Account,
+      amount : Nat,
+      onSuccess : (txIdx : Nat) -> async* WithdrawResponse,
+      onError : (err : ProcessWithdrawTransferError) -> async* WithdrawResponse,
+    ) : async* WithdrawResponse {
+      let callResult = await* processWithdrawTransfer(from_subaccount, to, amount);
 
       switch (callResult) {
-        case (#Ok txIdx) {
-          log(ownPrincipal, #withdraw({ to = to; amount = amount }));
-          #ok(txIdx, amount - fee(#withdrawal));
-        };
+        case (#Ok txIdx) { await* onSuccess(txIdx) };
         case (#Err(#BadFee { expected_fee })) {
           updateFee(expected_fee);
-          let retryResult = await* processWithdrawTransfer(to, amount);
+          let retryResult = await* processWithdrawTransfer(from_subaccount, to, amount);
           switch (retryResult) {
-            case (#Ok txIdx) {
+            case (#Ok txIdx) { await* onSuccess(txIdx) };
+            case (#Err err) { await* onError(err) };
+          };
+        };
+        case (#Err err) { await* onError(err) };
```

Comment:
I think there are problems with retrying here at this level in the BadFee case. First of all, it may not always work. For example, in the second call to `handleWithdrawal` (the second hop of the transfers) if the fee increased there won't be enough funds in the deposit account and we have to start over from the beginning.

And more abstractly, I don't think that code on this level should make any assumption on what `updateFee(expected_fee)` can change and what it cannot change. For example, it could change `fee(#withdrawal)` or the minimum and then I think we should start over again from the beginning instead of retrying a single step.